### DISCUSSION TITLE
Enhanced Machine Learning Dataset Support in CKAN

### DISCUSSION THREAD

**Discussion by anuveyatsu** [ckan/ckan#8414] [Ideas]
What people think about adding a support for https://github.com/mlcommons/croissant which would make ML datasets on CKAN instances more discoverable (and interoperable)? It could also promote CKAN as an option for organizations interested in hosting "ML-ready" datasets 🤔

I could imagine this being added to [DCAT](https://github.com/ckan/ckanext-dcat) extension or alternatively, as a standalone extension for CKAN. 

Please, post any thoughts on this below.

**Comment by pdelboca** [comment]
I think a `ckanext-croissant` is not only an amazing name for an extension, but also a good idea.

I'm +1 on extensions that map CKAN's metadata to other standards.

**Comment by benjelloun** [comment]
Croissant format author here. I support this idea :)

Seems like a simple way to add this support is to extend [ckanext/dcat/profiles/schemaorg.py](https://github.com/ckan/ckanext-dcat/blob/master/ckanext/dcat/profiles/schemaorg.py) with the additional properties defined by Croissant that can be extracted from existing CKAN metadata:
- Resources (e.g, Files, archive contents), which are described in more detail in Croissant than in schema.org
- Schemas of tabular files, which are captured in Croissant via RecordSet definitions. 

To help search engines index the datasets, Croissant metadata can replace the schema.org metadata currently embedded in the HEAD of the dataset Web pages. This is what ML dataset repositories like HuggingFace and Kaggle are doing.

**Comment by Reikyo** [comment]
Hi @pdelboca, I've been working with @benjelloun in recent weeks to learn about and modify CKAN for Croissant use, in particular focusing on the DCAT extension. I have a new profile, croissant.py, based on the existing schemaorg.py, and also a scheming schema for users that want to commit to a custom Croissant UI for their dataset and resource details. The latter is still up for discussion, and the former is the main goal at this point, to ensure that users can provide Croissant-enabled dataset search info by modifying their read_base.html as described [here](https://docs.ckan.org/projects/ckanext-dcat/en/latest/google-dataset-search/).

Can you tell me the process to enable new contributors to submit PRs to CKAN repos? And should they be based on master or a numbered release? Many thanks for your help.

**Reply by Reikyo** [reply]
@pdelboca Just bumping this comment from last week. We would really like to get these changes submitted as a PR soon in order for you and your team to check out.

**Comment by pdelboca** [comment]
Hello @Reikyo ! I'm happy to hear that you are working on this!

As I mentioned in my comment, the way to go is for you to create a new `ckanext-croissant` extension that implements the field mappings and provides an API endpoint to it (similar to what `ckanext-dcat` do with DCAT). I don't think it is a good idea to integrate it to the `ckanext-dcat` extension since the goal of the extension is compatibility with DCAT and not with multiple metadata standards. I you have strong arguments about it please open an issue in the extension itself to discuss it there.

What I propose is for you to create a `ckanext-croissant` extension that people can install and add it to plugins: https://docs.ckan.org/en/2.11/extensions/tutorial.html

Does it makes sense? The good thing of CKAN is that the Plugin based systems allows everyone to develop and share their improvements to CKAN.

**Reply by pdelboca** [reply]
Regarding the repository: You should create the extension in your own organization's Github Repository and share it with the community. The CKAN organization is reserved for core plugins that are maintained by the Tech Team of CKAN. Lots of important extensions are develop and mainteined by other organizations or companies, so they get plenty of visibility in the community.

**Reply by benjelloun** [reply]
Hi @pdelboca, I'd like to push back a bit on creating a separate extension for Croissant, for the following reasons:

1) schema.org/Dataset is already a separate vocabulary that is supported by the ckanext-dcat extension. Croissant is currently an extension of schema.org/Dataset, and a closer integration between Croissant and DCAT may make sense in the near future.

2) Croissant should replace schema.org/Dataset as the embedded mark-up in Web pages, to enable indexing by Croissant-aware tools like Google Dataset Search. This is difficult to achieve with two separate extensions.

3) Creating a separate extension also makes adoption a bit more difficult, as users will need to install and configure it for their repositories to have that functionality.

Do these make sense to you?

Best,
Omar

**Reply by pdelboca** [reply]
Hello @benjelloun ! 

I'm not the official maintainer of `ckanext-dcat` so I might be lacking some context here so I'll cc the @ckan/core to hear their opinions. Regarding your points:

1. I do not see the benefit of mixing DCAT with Croissant in the same extension. What do you mean with "a closer integration between Croissant and DCAT may make sense in the near future"? What are your plans for that integration? DCAT is a Data Catalog Vocabulary, while Croissant works at a Dataset level.

2. This means that we should replace the DCAT Dataset with Croissant definition?

3. Yes, it will make the adoption a bit more difficult, but the scope, maintainability and the governance of the codebase will be more clear. The aim of the plugin ecosystem is to have clear separation of concerns and better governance of codebases. Mixing DCAT with Croissant seems to break this.

I'm might be missing some context but unless there is a strong integration with DCAT and Croissant at project levels (not at CKAN implementation level) I don't think it will be a good idea to mix them in ckanext-dcat.

Let's hear about the rest of @ckan/core :)

**Reply by Reikyo** [reply]
Hi @pdelboca, many thanks for the comments, let's see what can be done here. I'll add some extra info in case I wasn't clear about the current status before.

1. I think Omar was referring to the fact that there is already a schemaorg.py profile in ckanext-dcat, and we are proposing a similar croissant.py profile to sit alongside it (in fact, the former is the basis for the latter). Such custom profiles are discussed in the extension's documentation [here](https://docs.ckan.org/projects/ckanext-dcat/en/latest/profiles/#writing-custom-profiles), which will enable a new RDF endpoint customisation as discussed [here](https://docs.ckan.org/projects/ckanext-dcat/en/latest/endpoints/#dataset-endpoints). Seeing as custom profiles can be made anyway, it would be good to share what we hope will be a popular option as standard.

2. By default, ckanext-dcat embeds JSONLD in the dataset page based on the schemaorg.py profile, not on one of the DCAT profiles, as discussed [here](https://docs.ckan.org/projects/ckanext-dcat/en/latest/google-dataset-search/). This is for Google dataset search, and Omar is the expert on that and can provide further insights.

To allow the adjustments to work as intended, the new profile does require tweaks to some other ckanext-dcat files just to register its presence and extra namespaces, in the exact same way that existing profiles do, nothing else. All logic is driven by the rest of the standard ckanext-dcat files. I did explore making a dedicated ckanext-croissant extension, but the existing features and logic of ckanext-dcat already supply exactly what's needed, and the changes are as according to how the extension authors propose that it can be modified (i.e. via new profiles and schemas). It would be a shame to simply fork ckanext-dcat to be sure of the full logic (which is complex), and make these relatively minor adjustments to create a new extension. Alternatively, inheriting from ckanext-dcat would pose a number of other development challenges.

**Comment by amercader** [comment]
For people interested in this, just a quick note to say that the just released [ckanext-dcat 2.3.0](http://127.0.0.1:8000/projects/ckanext-dcat/changelog/#v230-2025-02-25) includes initial support for Croissant ML, allowing to expose Datasets, Resources and DataStore data as the [Croissant format](https://docs.mlcommons.org/croissant/docs/croissant-spec.html) Datasets, Resources and RecordSets.

To learn more check the documentation, that includes info about installation (just enable the `croissant` plugin), available schemas, endpoints, etc:

https://docs.ckan.org/projects/ckanext-dcat/en/latest/croissant/

This is an initial implementation and there are things that might need ironing out, if someone is interested in testing and provide feedback it that would be very appreciated.
